store.go

```diff
@@ -1151,6 +1179,34 @@ func (s *store) CreateContainer(id string, names []string, image, layer, metadat
 
 	var imageTopLayer *Layer
 	imageID := ""
+
+	if options.AutoUserNs || options.UIDMap != nil || options.GIDMap != nil {
+		// Prevent multiple instances to retrieve the same range when AutoUserNs
+		// are used.
+		// It doesn't prevent containers that specify an explicit mapping to overlap
+		// with AutoUserNs.
+		s.usernsLock.Lock()
+		defer s.usernsLock.Unlock()
+	}
+	if options.AutoUserNs {
+		var err error
+		size := uint32(0)
+		if val, found := options.AutoUserNsOpts["size"]; found {
+			s, err := strconv.Atoi(val)
+			if err != nil {
+				return nil, err
+			}
+			if s < 0 {
+				return nil, errors.Errorf("invalid size specified for the user namespace %q", val)
+			}
+			size = uint32(s)
+		}
+		options.UIDMap, options.GIDMap, err = s.getAutoUserNS(size, image)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Could we turn this code into a anonymous function/lambda? This would reduce the critical section and hence the time we keep the lock (and ultimately speed it up a bit more).

Reply:
```golang
uidMap, gidMap, err := func () ... {
   lock()
   defer unlock()
}

Reply:
we need to serialize the entire function.  We can release the lock only after the container is created in the store to prevent another process to pick the same range.